compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -555,16 +556,21 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn reserve_and_set_fn_alloc(self, instance: Instance<'tcx>) -> AllocId {
         // Functions cannot be identified by pointers, as asm-equal functions can get deduplicated
         // by the linker (we set the "unnamed_addr" attribute for LLVM) and functions can be
-        // duplicated across crates.
-        // We thus generate a new `AllocId` for every mention of a function. This means that
-        // `main as fn() == main as fn()` is false, while `let x = main as fn(); x == x` is true.
-        // However, formatting code relies on function identity (see #58320), so we only do
-        // this for generic functions. Lifetime parameters are ignored.
+        // duplicated across crates. We thus generate a new `AllocId` for every mention of a
+        // function. This means that `main as fn() == main as fn()` is false, while `let x = main as
+        // fn(); x == x` is true. However, formatting code relies on function identity (see #58320)
+        // -- that's likely broken, but for now we have to support it to make Miri work. So we
+        // identify whether codegen will actually emit duplicate functions. It does that when they
+        // have non-lifetime generics, or when they can be inlined.
         let is_generic = instance
             .args
             .into_iter()
             .any(|kind| !matches!(kind.unpack(), GenericArgKind::Lifetime(_)));
-        if is_generic {
+        let can_be_inlined = match self.codegen_fn_attrs(instance.def_id()).inline {
+            InlineAttr::Never => false,
+            _ => true,
+        };
```

Comment:
Generic functions are already handled above.

The direction we need is that if the item can be cross-crate inlined then we should give it a fresh address each time here, i.e. `is_generic || can_be_inlined` should be true. It's okay if we give it a fresh address too often. I am not sure if that property holds with my current definition, it's hard to say as everything is spread across so many parts of rustc.

Cc @saethlin

Reply:
It is and was entirely unclear to me if any parts of rustc _rely_ on certain Instances getting LocalCopy or GloballyShared codegen. I've been meaning to try to clean this logic up at some point.

Cross-crate inlining should not be orthogonal from the attribute. It should return true for anything that is `#[inline]` or `#[inline(always)]`, and use heuristics for all other cases.

I do not understand why we permit `#[no_mangle]` with `#[inline(always)]`. That seems like an incoherent request to me. `#[no_mangle]` is widely understood as requesting that the function be lowered as a single external symbol that can be linked against, and lowering it as LocalCopy directly contradicts that. We should probably produce a hard error for that attribute combination.

Reply:
`#[no_mangle]` + `#[inline(always)]` to me seems like a request to export a single copy for consumption by C code, while still codegening a separate local copy for each use from Rust with the request to inline it where possible.

Reply:
FWIW this is all off-topic for this PR. The key thign this PR relies upon is that if a function is non-generic (only lifetimes) and `inline(never)` then it will never have more than one copy across the entire crate graph.

Reply:
And even that is a rather soft requirement -- the LLVM attribute we set makes it so that functions can still be (de)duplicated. It's just that making all functions duplicated in Miri makes backtraces look worse so it's kind of not worth it...

Ideally we'd have an attribute that makes a function guaranteed to have a unique address, and then we suppress the unnamed_addr attribute. Then we can decorate the backtrace functions with that.

But until then this PR at least surfaces more possible address mismatches than before. I am just a bit worried about generating an endless amount of `AllocId` if a function is cast to a pointer *a lot*...

Reply:
> I am just a bit worried about generating an endless amount of `AllocId` if a function is cast to a pointer _a lot_...

Shouldn't the GC prevent any issues with this?

Reply:
The GC doesn't remove allocations, I don't think?

Reply:
Right, it doesn't. I'll file a Miri issue as a possible extension, but for now this program:
```rust
fn main() {
    loop {
        oof as usize;
    }
}

fn oof() {}
```
seems to take a minute for its memory usage to grow to 1 GB. That is definitely not desirable but in general I think programs that do that number of casts will have other runtime or memory usage issues.